rust/shard_client/src/local_shard_client.rs

```diff
@@ -1,22 +1,124 @@
 use async_trait::async_trait;
+use bincode::Options;
+use cas::safeio::write_all_file_safe;
 use cas_client::{Client, LocalClient};
+use itertools::Itertools;
+use mdb_shard::constants::MDB_SHARD_GLOBAL_DEDUP_CHUNK_MODULES;
 use mdb_shard::error::{MDBShardError, Result};
 use mdb_shard::file_structs::MDBFileInfo;
-use mdb_shard::MDBShardFile;
+use mdb_shard::shard_dedup_probe::ShardDedupProber;
 use mdb_shard::{shard_file_reconstructor::FileReconstructor, ShardFileManager};
+use mdb_shard::{MDBShardFile, MDBShardInfo};
+use merkledb::aggregate_hashes::with_salt;
 use merklehash::MerkleHash;
-use std::io::Cursor;
-use std::path::PathBuf;
+use std::collections::HashMap;
+use std::io::{BufReader, Cursor, Read, Write};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::sync::RwLock;
 
 use crate::{RegistrationClient, ShardClientInterface};
 
+struct LocalGlobalDedupTable {
+    table: Arc<RwLock<HashMap<String, String>>>, // map of prefix/chunk_hash -> prefix/shard_hash
```

Comment:
Is there ever a situation where we would have different prefixes?   If so, let's instead have a HashMap<Prefix, HashMap<MerkleHash, MerkleHash> >.  The reason here is that it would be much faster to use the regular MerkleHash as the type, as each String will require a memory allocation whereas the MerkleHash type is fixed.

Reply:
Chatted offline -- going to use a different technique here, but still use the prefix in the key and value.

Reply:
Finally implemented using a leveldb. Added tests for multithreaded read/write and multi-instance read/write.